tests/test_DGEList.py

```diff
@@ -213,7 +213,23 @@ def test_rpkm():
     # RPKM=numReads / (geneLength / 1000 * totalNumReads / 1, 000, 000)
     assert rpm_dge.counts[0][0] == (first_pos / ((gene_len / 1e3) * (col_sum[0] / 1e6)))
 
+def test_tpm():
+    dge_list = DGEList(filename=str(get_dataset_path(TEST_DATASET_NPZ)))
+    icd = CanonicalDataStore(
+        get_dataset_path(TEST_GENE_SET_DATA), get_dataset_path(TEST_GENE_SYMBOLS)
+    )
+    first_pos = dge_list.counts[0][0]
+    first_gene = dge_list.genes[0]
 
+    col_sum = np.sum(dge_list.counts, axis=0)
+    assert isinstance(first_pos, np.integer)
+    tpm_dge = dge_list.tpm(icd)
+    ensg_gene = icd.pick_gene_id(icd.get_genes_from_symbol(first_gene))
```

Comment:
This is similar to how I designed the the test_tpm in the first run. Here is the comment by clintval to that:
https://github.com/r-bioinformatics/edgePy/pull/72/commits/2276d19210c76c811d99a8864d42431646cfbf10#r224543989

That's the reason why I chose a very simple table that is manually calculatable

Reply:
I guess I'm a bit confused. Is this in reference to the comments about using random values? This test shouldn't be using anything random; it should be pulling the first entry from the test data, similar to what the RPKM test is doing.

Reply:
The link I posted is not working as I expected! I'll copy the comment here:
clintval: "Not bad, you are dynamically creating an expected result by copying some code from the main application.

This makes it hard to track down bugs with the logic since the logic exists in two places.

Norm. methods should probably be tested with a very small, hand-designed test case so you can explicitly write out all inputs, and all outputs."

This is why I didn't implement the test_tpm function like the test_rpkm function in the end. As a consequence also test_rpkm shouldn't be implemented as is at the moment..

Reply:
Okay, that makes much more sense. In that case, it may make sense to us apfejes' RPKM function sig and your test function as standards. It seems reasonable but probably worth discussing (unless it's already been discussed before I arrived).

Would it be worth it to possibly create a minimal test data set and run it through edgeR to match the edgeR outputs to ours? That's probably a larger discussion, but known and verified 3rd-party outputs may make sense for test cases.

Reply:
Consider too, an additional layer of testing as property-based for these mathematical calculations:

https://hypothesis.readthedocs.io/en/latest/data.html